CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+
+**Recursion**
+
+Code that implements recursive algorithms must be supported by a justification
+showing that the execution time and memory requirements are bounded and
+compatible with the target hardware and software architecture.
+
+**Unused Return Values**
+
+The return value of a non-void function must be explicitly cast to (void)
+using a C-style cast if the value is not used by the calling function.
+
+[I believe the purpose here is to show that anyone ignoring the return value 
+of a function is doing so intentionally, but I'm considering pulling it anyway
+because it's a little weird.  Thoughts?]
+
+**Order of Operations**
+
+Expressions using two or more operators of different precedence must use
+parentheses to clarify the binding of operands to operators when the parsing
+order is not intuitive.
+
+**Types**
+
+For integer types in C and C++ code, use int, unsigned int, plain char and bool
+when possible, and don't use precise-width integer types in the absence of a
+specific sizing requirement.
+
+In other words, if a specific size is needed, a precise-width integer type
+(e.g., uint16_t) from `<stdint.h>` or equivalent must be used, and the following
+fundamental types must not be used: signed char, unsigned char, short, unsigned
+short, long, unsigned long, long long, unsigned long long. Exception: These
+types may be used to define precise-width types in `<stdint.h>` or equivalent.
+
+Only these standard C++ fundamental numeric types should be used:
+
+* `bool`
+* `char` (7-bit ASCII character)
+* `int` (32 bits, signed)
+* `unsigned` (32 bits, unsigned)
+* `size_t` (pointer size, 32 bits, unsigned)
+* `ptrdiff_t` (pointer difference size, 32 bits, signed)
+* `int8_t`
+* `int16_t`
+* `int32_t`
+* `int64_t`
+* `uint8_t`
+* `uint16_t`
+* `uint32_t`
+* `uint64_t`
+* `float`
+* `double`
+
+The C++ standard only guarantees the size of `int` and `unsigned` are at
+least 16 bits, but we assume they are 32 bits, as is common for 32-bit
+platforms.
+
+Signed values (`int`) should be prefered for representing quantities.  Unsigned
+values (`unsigned`) should not be used to "guarantee" that a function parameter
+or other variable is non-negative.  Assertions should be used instead.
+
+The C++ standard also does not specify the signedness of `char`, so is up to
+the compiler implementer to choose or make it a configurable option.  Do not
+rely on that choice.  Instead, use `int8_t`/`uint8_t` and use `char` only if
+referring to a character in a plain C-style string.
+
+Do not use `long double`.
+
+
+**Mutable**
+
+Using the mutable keyword requires justification.
+
+**Exceptions**
+
+*C++ exceptions must not be used.*
+
+[This was more of a hardware limitation for UASC, but it may make sense for us as 
+
+well. 
+Read up on the
+<a href="https://en.wikipedia.org/wiki/Design_by_contract">Design By Contract</a>
+programming style (possibly to be included as a separate document).]
+
```

Comment:
Not agreeing with you yet. Does this rule out sections of the std not marked noexcept? If this is identified as a priority then it needs to be reworded as end user code or our code. I don't think the ros libraries make liberal use of noexcept so that would be impossible to enforce.

Reply:
True.  Perhaps the only reason to avoid exceptions for us is that they, rarely, can leave your OS in an unexpected state after having been handled.  <a href="https://randomascii.wordpress.com/2012/07/05/when-even-crashing-doesnt-work/">See this article for details</a> (and quite a bit of other interesting tidbits).

Granted, the problem he describes is with Windows 64-bit, particularly, but it's strange enough to make me wary of exceptions in general.